src/usdb_syncer/notes_parser.py

```diff
@@ -51,6 +51,11 @@ def parse(cls, value: str) -> Note:
             pitch = int(match.group(4))
         except ValueError as err:
             raise NotesParseError(f"invalid note: '{value}'") from err
+        if kind != NoteKind.FREESTYLE:
```

Comment:
What about the opposite case? Shouldn't we keep deleting freestyle notes without text?

Reply:
Good point. I guess we should.

Reply:
I also just stumbled upon a note of the form `: 1 2 3` (missing space after duration and missing text) which should also be handled as an empty note that needs a `~`.

Reply:
Yes, as far as I can see, there is no reason to keep empty freestyle notes.

Oh, right. The last space also has to be made optional.

Reply:
So then `regex = re.compile(r"(:|\*|F|R|G):? +(-?\d+) +(\d+) +(-?\d+) ?(.*)?")` (also the last capture group optional?)

Reply:
The last `?` is redundant, because `(.*)` will always match (the empty string).
I would say, if there are lyrics they must be separated by a space. The above regex would match `: 1 1 1foo` for example.
That would mean something like this: `(:|\*|F|R|G):? +(-?\d+) +(\d+) +(-?\d+)(?: (.*))?`. But note that now `match.group(4)` can be `None`.

Reply:
On a different note, `parse()` was an `__init__()` previously, so I couldn't return `Note | None` and had to raise an exception in case the note could not be parsed. If it makes things easier, you could remove the exception, move the logging to `Note.parse()` and return `None` on a parse failure.

Reply:
Yes, I agree that if there are lyrics, they must be separated by a space and we should reject `: 1 1 1foo` as invalid. It should only work for empty notes if there is no space after the pitch (which should then get ~ as text).

Thus:
`: 1 1 1` --> `: 1 1 1 ~`
`: 1 1 1 `--> `: 1 1 1 ~`
`: 1 1 1  ` --> `: 1 1 1 ~ `

Reply:
Where did your other remarks go?

Reply:
Somehow the other comment ended up in a different thread. 🤔

Reply:
> On a different note, `parse()` was an `__init__()` previously, so I couldn't return `Note | None` and had to raise an exception in case the note could not be parsed. If it makes things easier, you could remove the exception, move the logging to `Note.parse()` and return `None` on a parse failure.

Not sure if I fully understand this. Let’s exchange on this in a call some time.